File: NetworkingDotNet/NetworkPacketManipulator.cs
/*
TCP Networking Library (.Net Version)

Author: Recep ALTIN 
*/
using System;
using System.Collections.Generic;

namespace NetworkingDotNet
{
    public class NetworkPacketManipulator
    {
        public static bool Split(NetworkPacket _np, ref List<NetworkPacket> _nps)
        {
            _np.ToBuffer(); // force override to fill buffer

            if (_np.Data.Count > NetworkingDotNetGlobals.MaximumDataSize)
            {
                int i = 0;
                uint partOrder = 0;
                uint partCount = (uint)(Math.Ceiling((double)_np.Data.Count / (double)NetworkingDotNetGlobals.MaximumDataSize));
                List<byte> ba = new List<byte>();

                while (i < _np.Data.Count)
                {
                    ba.Add(_np.Data[i]);
                    if (++i == ((partOrder + 1) * NetworkingDotNetGlobals.MaximumDataSize) || i == _np.Data.Count)
                    {
                        NetworkPacket np = new NetworkPacket(i == _np.Data.Count ? NetworkPacketType.NPT_PARTIAL_DATA_END : NetworkPacketType.NPT_PARTIAL_DATA)
                        {
                            Id = _np.Id,
                            TypeId = _np.TypeId,
                            PartOrder = partOrder++,
                            PartCount = partCount,
                            Data = new List<byte>(ba)
                        };
                        _nps.Add(np);
                        ba.Clear();
                    }
                }
            }
            else
            {
                _nps.Add(new NetworkPacket(_np));
            }

            return true;
        }

        public static bool Join(List<NetworkPacket> _nps, ref NetworkPacket _np)
        {
            if (_nps.Count > 0)
            {
                List<byte> data = new List<byte>();

                for (int i = 0; i < _nps.Count; i++)
                {
                    NetworkPacket p = _nps[i];

                    if (i != p.PartOrder)
                    {
                        return false;
                    }

                    data.AddRange(p.Data);

                    if (p.PacketType == NetworkPacketType.NPT_PARTIAL_DATA_END)
                    {
                        _np.Id = p.Id;
                        _np.TypeId = p.TypeId;
                        _np.Data = new List<byte>(data);
                    }
                }
                return true;
            }

            return false;
        }
    }
}

File: Networking/networkpacketmanipulator.cpp
#include "networkpacketmanipulator.h"
#include <QtMath>

using namespace Networking;
NetworkPacketManipulator::NetworkPacketManipulator(void)
{
}

NetworkPacketManipulator::~NetworkPacketManipulator(void)
{
}

bool NetworkPacketManipulator::split(Networking::NetworkPacket & _np, QList<NetworkPacket> * _nps)
{
	_np.toBuffer(); // force override to fill buffer

	QByteArray & data = _np.data();
	if (data.size() > NetworkingGlobals::FixedBufferSize)
	{
		int i = 0;
		uint partOrder = 0;
		uint partCount = (uint)(qCeil((double)data.size() / (double)NetworkingGlobals::MaximumDataSize));
		quint32 id = _np.id();
		quint32 type_id = _np.typeId();

		QByteArray ba;
		while (i < data.size())
		{
			ba.append(data.at(i));
			if (++i == ((partOrder + 1) * NetworkingGlobals::MaximumDataSize) || i == data.size())
			{
				NetworkPacket np(i == data.size() ? NPT_PARTIAL_DATA_END : NPT_PARTIAL_DATA);
				np.setId(id);
				np.setTypeId(type_id);
				np.setPartOrder(partOrder++);
				np.setPartCount(partCount);
				np.setData(ba);
				_nps->append(np);
				ba.clear();
			}
		}
	}
	else
		_nps->append(_np);

	return true;
}

bool NetworkPacketManipulator::join(QList<NetworkPacket> & _nps, Networking::NetworkPacket * _np)
{
	if (_nps.size() > 0)
	{
		QByteArray data;
		int ord = 0;

		for (quint32 i = 0; i < _nps.count(); i++)
		{
			NetworkPacket * p = const_cast<NetworkPacket *>(&_nps.at(i));

			if (i != p->partOrder())
				return false;

			data.append(p->data());

			if (p->packetType() == NPT_PARTIAL_DATA_END)
			{
				_np->setId(p->id());
				_np->setTypeId(p->typeId());
				_np->setData(data);
			}
		}
		return true;
	}

	return false;
}
